primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -652,6 +655,164 @@ where
+        let comm = Self::load_commitment_to_host::<C>(msm_result_on_gpu, stream)?;
+        Ok(comm)
+    }
+
+    /// load SRS for prover onto GPU once, and reuse for future poly-commit of
+    /// degree<=`supported_degree`
+    pub fn load_prover_param_to_gpu<'srs, C>(
+        prover_param: impl Borrow<UnivariateProverParam<E>>,
+        supported_degree: usize,
+    ) -> Result<HostOrDeviceSlice<'srs, IcicleAffine<C>>, PCSError>
+    where
+        C: IcicleCurve + MSM<C>,
+        C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+        C::BaseField: ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+        E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+    {
+        let prover_param = prover_param.borrow();
+        if supported_degree > prover_param.powers_of_g.len() - 1 {
+            return Err(PCSError::InvalidParameters(format!(
+                "supported degree {} is larger than allowed {}",
+                supported_degree,
+                prover_param.powers_of_g.len() - 1
+            )));
+        }
+
+        let mut bases_on_device =
+            HostOrDeviceSlice::<'_, IcicleAffine<C>>::cuda_malloc(supported_degree + 1)?;
+
+        #[cfg(feature = "kzg-print-trace")]
+        let conv_time = start_timer!(|| "Type Conversion: ark->ICICLE: Group");
+        let bases: Vec<IcicleAffine<C>> =
+            parallelizable_slice_iter(&prover_param.powers_of_g[..supported_degree + 1])
+                .map(|&p| IcicleAffine::<C>::from_ark(p))
+                .collect();
+        #[cfg(feature = "kzg-print-trace")]
+        end_timer!(conv_time);
+
+        #[cfg(feature = "kzg-print-trace")]
+        let load_time = start_timer!(|| "Load group elements: CPU->GPU");
+        bases_on_device.copy_from_host(&bases)?;
+        #[cfg(feature = "kzg-print-trace")]
+        end_timer!(load_time);
+
+        Ok(bases_on_device)
+    }
+
+    /// Load polynomial's coefficients onto GPU, preparing for poly-commit on
+    /// GPU later
+    pub fn load_poly_to_gpu<'poly, C>(
+        poly: &DensePolynomial<E::ScalarField>,
+    ) -> Result<HostOrDeviceSlice<'poly, C::ScalarField>, PCSError>
+    where
+        C: IcicleCurve + MSM<C>,
+        C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+        C::BaseField: ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+        E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+    {
+        let size = poly.degree() + 1;
+        let mut scalars_on_device = HostOrDeviceSlice::<'_, C::ScalarField>::cuda_malloc(size)?;
+
+        #[cfg(feature = "kzg-print-trace")]
+        let conv_time = start_timer!(|| "Type Conversion: ark->ICICLE: Scalar");
+        let scalars: Vec<C::ScalarField> = parallelizable_slice_iter(&poly.coeffs()[..size])
+            .map(|&s| C::ScalarField::from_ark(s))
```

Comment:
use `from_bytes_le(s.into_bigint().to_bytes_le()` instead.

Reply:
If we hardcode these field types, maybe we should use `from_limbs` directly.

Reply:
If we do not use generics, we may directly use, e.g. `icicle_bn254::curve::ScalarField::from(ark_bn254::Fr::from(1u64).0.0)`

Reply:
Even faster conversion if we avoid using generics: https://github.com/EspressoSystems/jellyfish/blob/cl/type-conversion/utilities/src/conversion.rs#L310-L339

Reply:
[ultimate solution](https://github.com/EspressoSystems/jellyfish/blob/9898c73931ca7d5819da99aadb9dcb49381c95ec/utilities/src/conversion.rs#L311-L365)